text/0000-cargo-workspace-deduplicate.md

```diff
@@ -0,0 +1,633 @@
+we can specify a few directives. Naturally that lends itself to quite a lot of
+possible bikeshedding! Virtually all of the aspects of the proposal that modify
+`Cargo.toml` can be tweaked in various ways such as names used or where they're
+placed. In any case discussion about compelling alternatives is always
+encouraged!
+
+## Not including metadata by default
+
+One possible change to the proposal as well is to not include missing metadata
+from `[package]` by default from `[workspace]`, if defined. For example an
+explicit opt-in could be required:
+
+```toml
+[package]
+repository = { workspace = true }
+```
+
+or you could blanket opt-in everything
+
+```toml
+[package]
+inherit-workspace-metadata = true
+```
+
+These options feel a bit more verbose than necessary, though, and not
+necessarily aligned with Cargo's existing manifest idioms where it infers things
+like `[[bin.name]]` or lists of tests and such.
+
+## Motivating issues
+
+Duplication throughout workspaces has been a thorn in Cargo's since practically
+since the inception of workspaces. Naturally there's quite a few bugs filed on
+Cargo's issue tracker about this which provide some context for why make a
+proposal at all as well as how to design this proposal.
+
+* [#3931] - updating the version of a crate in a workspace means lots of edits
+* [#7552] - crates may differ slightly in versions required from crates.io
+* [#7964] - current idioms push users towards usage of `[patch]` which breaks
+  git dependencies
+* [#5471] - an issue about shared dependencies in a workspace
+* [#6126] - an issue where `[patch]` tables are used seemingly to make it easier
+  to specify dependencies in a workspace, but having everything in
+  `[workspace.dependencies]` makes it smaller to specify.
+* [#6828] - an issue about inheriting workspace attributes
+
+[#3931]: https://github.com/rust-lang/cargo/issues/3931
+[#7552]: https://github.com/rust-lang/cargo/issues/7552
+[#7964]: https://github.com/rust-lang/cargo/issues/7964
+[#5471]: https://github.com/rust-lang/cargo/issues/5471
+[#6126]: https://github.com/rust-lang/cargo/issues/6126
+[#6828]: https://github.com/rust-lang/cargo/issues/6828
+
+## Full templating language
+
+One sort of far-out-there alternative we could go for is to be far more
+ambitious and make our own sort of "templating language" on top of TOML. This
+would arguably be much more flexible than the limited amount of deduplication
+proposed here, but you could imagine things like:
+
+```toml
+[package]
+name = "foo"
+version = "1.{workspace.vars.minor}.0"
+
+[dependencies]
+bar = "{workspace.dependencies.bar}"
+baz = { version = "1", features = "{workspace.vars.baz_features}" }
+```
+
+or "insert your own idea for how we can go all out" here. In general though I
+think there's a lot to be gained from the simplicity of TOML and prioritizing
+other tools reading Cargo manifests, so we may not want to go full-blown
+templating language just yet.
```

Comment:
It seems like the value of inheriting the `documentation` field in particular will be limited without the ability to template-expand the crate name into the URL, right?

Reply:
I agree yeah that the usefulness in all cases is pretty limited, but that's somewhat intentional as well. I suspect this will largely be used for homepage/license/etc information, but there wasn't much reason to leave out keys like `documentation` as well. It may be useful for 2-workspace crates though where one's a macro internals and the other is the real crate, where both effectively are documented through the same url.